### ISSUE TITLE
docker: Running minio server in docker container fail to work with symlink PATH

### ISSUE THREAD

**Issue by minherz** [minio/minio#3742]
When running minio server on Windows docker host in a windows platform container and providing it a path which is mounted volume to external folder, the "slice bounds out of range" error is reported.
The issue does not happened when the mounted volume is provided as `--config-dir` parameter.

## Expected Behavior
Providing mounted volume in Windows docker container expected to work.

## Current Behavior
Attempt to start minio server on the Windows docker container when server path points to mounted folder results in the following error. If only configuration directory is defined as mounted folder but the server path is local for the container, the server starts as usual.
```
panic: runtime error: slice bounds out of range

goroutine 1 [running]:
panic(0xbb2a60, 0xc042004040)
        /home/builder/.gimme/versions/go1.7.4.linux.amd64/src/runtime/panic.go:500 +0x1af
syscall.Readlink(0xc0423869f0, 0xd, 0xc042382500, 0x80, 0x80, 0x0, 0x0, 0x0)
        /home/builder/.gimme/versions/go1.7.4.linux.amd64/src/syscall/syscall_windows.go:1031 +0x3a9
os.Readlink(0xc0423869f0, 0xd, 0x108db40, 0xc04238e0c0, 0x0, 0x0)
        /home/builder/.gimme/versions/go1.7.4.linux.amd64/src/os/file_posix.go:21 +0x94
os.Stat(0xc0423869f0, 0xd, 0xc0423869f0, 0xd, 0x0, 0x0)
        /home/builder/.gimme/versions/go1.7.4.linux.amd64/src/os/stat_windows.go:72 +0x9e
github.com/minio/minio/cmd.newFSObjectLayer(0xc0423869e0, 0x9, 0xc0423869c0, 0x7, 0x0, 0x0)
        /home/builder/mygo/src/github.com/minio/minio/cmd/fs-v1.go:99 +0xce
github.com/minio/minio/cmd.newObjectLayer(0xc042276569, 0x5, 0xc04206a390, 0x1, 0x1, 0x1, 0x1, 0x0, 0x0)
        /home/builder/mygo/src/github.com/minio/minio/cmd/server-main.go:493 +0x32e
github.com/minio/minio/cmd.serverMain(0xc0420e01e0)
        /home/builder/mygo/src/github.com/minio/minio/cmd/server-main.go:465 +0x959
github.com/minio/minio/vendor/github.com/minio/cli.Command.Run(0xc96a1c, 0x6, 0x0, 0x0, 0x0, 0x0, 0x0, 0xcac1d6, 0x1c, 0x0, ...)
        /home/builder/mygo/src/github.com/minio/minio/vendor/github.com/minio/cli/command.go:117 +0xbc8
github.com/minio/minio/vendor/github.com/minio/cli.(*App).Run(0xc04218f680, 0xc04206c050, 0x5, 0x5, 0x0, 0x0)
        /home/builder/mygo/src/github.com/minio/minio/vendor/github.com/minio/cli/app.go:186 +0x24d
github.com/minio/minio/vendor/github.com/minio/cli.(*App).RunAndExitOnError(0xc04218f680)
        /home/builder/mygo/src/github.com/minio/minio/vendor/github.com/minio/cli/app.go:197 +0x56
github.com/minio/minio/cmd.Main()
        /home/builder/mygo/src/github.com/minio/minio/cmd/main.go:227 +0x7a
main.main()
        /home/builder/mygo/src/github.com/minio/minio/main.go:28 +0x1b
```

## Possible Solution
Since it does not happened for configuration directory, it is possible that an expected behavior is not supported for Windows containers while it exists for Linux containers. In this case, the issue can be forwarded to [docker for windows](https://github.com/docker/for-win/issues/).

## Steps to Reproduce (for bugs)
1. Create a windows container using the following docker file (assume that minio.exe is located in the same folder where Dockerfile):
```
FROM microsoft/dotnet:nanoserver
LABEL minio_version="2017-01-25T03:14:52Z"

COPY minio.exe c:\\app\\

RUN mkdir c:\\volume;  mkdir c:\\.minio

VOLUME c:\\volume c:\\.minio

WORKDIR c:\\app\\

EXPOSE 9000

CMD ["CMD"]
```
2. Make the Dockerfile directory to be current directory
3. Run `docker build . -t minio-server` to build docker image
4. Run `docker run -it -p 9000:9000 -v d:\Work\.minio:c:\.minio -v d:\Work\Data:c:\volume minio-server` to run the container
5. In the prompt of the container type `minio.exe server -C c:\.minio c:\volume`

## Context
Creating minio container with the shared data folder is essential for distributed access and recovery of the object storage.

## Your Environment
* Minio Version used: 2017-01-25T03:14:52Z
* Operating System and version: Windows 10 x64 Enterprise Version 1607 (Build 14393.693)
* Docker for Windows version: Version 1.13.1-beta42 (10069) in Beta mode

**Comment by vadmeste** [comment]
@minherz, I tested the steps you provided with Windows Server 2016, and I've got the same exact error. I have this minimal code which generates also the same stack trace:

```
package main

import "fmt"
import "os"

func main() {
	path := "\\\\?\\c:\\volume" // \\?\c:\volume or c:\volume
	fmt.Printf("Stating %s\n", path)
	_, err := os.Stat(path)
	if err == nil {
		fmt.Printf("Works as expected.\n")
	}
}
```

Not sure if this is a golang or Docker bug (most likely the latter). But at least we know that Minio has nothing to do with this.

A painful workaround is to do the following:
$ Enter the container
$ Run `dir c:\` to find which directory `c:\volume` is linked to
$ Create a new symlink directory linked to the same directory targeted by `c:\volume`. It would be something like below:

`mklink /D c:\myvolume \\?\ContainerMappedDirectories\A36A3BE1-96D7-4B62-8AEA-51C36D23CD8E`

and then, you run Minio normally. `minio.exe server c:\myvolume`

Keep me updated.

**Comment by minherz** [comment]
Thanx. Since I do not have golang environment, can you please, prepare for me .exe file that does the work on a directory which is passed to it as a single argument? This way i can propogate it to docker guys.

**Comment by minherz** [comment]
@vadmeste, does it mean that treatment of the configuration directory does not use the problematic code functionality, since the problem happens _only_ when the data storage location is set to mounted volume? Having the configuration directory pointing to the mounted volume works.

**Comment by harshavardhana** [comment]
> @vadmeste, does it mean that treatment of the configuration directory does not use the problematic code functionality, since the problem happens only when the data storage location is set to mounted volume. Having the configuration directory pointing to the mounted volume works.

The issue can occur in certain situations the `syscall.Readlink` implementation in Go for windows relies on a specific path length with some UTF translation code. The issue seems here is that docker nanoserver has done some changes which is causing trouble here.  Most probably this is a bug in the nanoserver itself which has perhaps broken backward compatibility with other windows versions. 

We will send you an executable reproducing this issue consistently.

**Comment by vadmeste** [comment]
This is x64 windows binary https://slack-files.com/T3NG3BE8L-F45KHQJTT-e9cf313b78, 
Below is the correspondant go code:

```
package main

import "fmt"
import "os"

func main() {    
    if len(os.Args) != 2 {
        fmt.Println("USAGE:  sample.exe path")
        os.Exit(1)
    }
    path := os.Args[1]
    fmt.Printf("Call os.Stat(%s)\n", path)
    _, err := os.Stat(path)
    fmt.Printf("err = %v\n", err)
    
}
```

This is the line of crash https://github.com/golang/go/blob/6b36535cf382bce845dd2d272276e7ba350b0c6b/src/syscall/syscall_windows.go#L1031

**Comment by minherz** [comment]
Hello, the line of crash that you sent me does not match the behavior of the system and also does not match the "latest" version of the code. [This line](https://github.com/golang/go/blob/master/src/syscall/syscall_windows.go#L1031) is 1031 for the syscall_windows.go in the master branch. I wrote a dotnet app that calls `DeviceIOControl` WIN32 API and prints results. For the SYMLINKD folders it shows valid values where SubstituteName which golang uses is something like: `\ContainerMappedDirectories\A07E52BD-3B5B-44EB-8648-7DDB69D26682`.
It seems that golang code crashes on interpretation of the string and not on the invalid structure that DeviceIOControl returns.
I have added [this test program](https://slack-files.com/T3NG3BE8L-F46CVAYEN-8247786785) (requires .net core 1.1) just in case

**Comment by vadmeste** [comment]
> Hello, the line of crash that you sent me does not match the behavior of the system and also does not match the "latest" version of the code. This line is 1031 for the syscall_windows.go in the master branch.

@minherz, according to the stacktrace you pasted above, go 1.7.4 was used and the crash happened in syscall_windows.go L1031. The link that I pasted points to the version of syscall_windows.go when 1.7.4 was just released.

**Comment by vadmeste** [comment]
@minherz, this could be fixed by this commit https://github.com/golang/go/commit/e65bce7144dbced232df8842ef6825d7e45f094e. I'll send you a new Minio binary with the last go runtime.

**Comment by minherz** [comment]
Thank you. I still opened an issue in docker/for-win since the behavior seem to differ from one that is expected. It can be that there is some issue in the virtual FS that was created to support containers in Windows.

**Comment by harshavardhana** [comment]
Since we will be moving to newer go1.8 release you can expect this to be fixed in minio/minio:edge tag.

**Comment by vadmeste** [comment]
> Since we will be moving to newer go1.8 release you can expect this to be fixed in minio/minio:edge tag.

Yes, we don't have that crash anymore, but it seems there is another bug

```
C:\>sample.exe c:\volume
Call os.Stat(c:\volume)
err = GetFileAttributesEx \ContainerMappedDirectories\BB1FFC92-773D-43B0-BE5E-1E81AA053643: The system cannot find the path specified.
```

(without '\\\\?\')

```
C:\>sample.exe \\?\ContainerMappedDirectories\BB1FFC92-773D-43B0-BE5E-1E81AA053643
Call os.Stat(\\?\ContainerMappedDirectories\BB1FFC92-773D-43B0-BE5E-1E81AA053643)
err = <nil>
C:\>
```

@minherz wrote a dotnet code to test without golang.

**Comment by harshavardhana** [comment]
Oh so looks like UNC paths work fine which should be okay with Minio, since we only use UNC paths. It is strange to see simple style doesn't work anymore. 

Can you paste the dotnet code? @vadmeste

**Comment by minherz** [comment]
@harshavardhana , it is not the UNC path but a symlinkd created by docker (for windows) when running minio using these symlinkd as path to data.
the code is relatively lengthy, how do you want to post it?

**Comment by harshavardhana** [comment]
> @harshavardhana , it is not the UNC path but a symlinkd created by docker (for windows) when running minio using these symlinkd as path to data.

It is strange to see that UNC path doesn't trigger the same issue, at this point we can be sure that this is a nanoserver bug. 

> the code is relatively lengthy, how do you want to post it?

Then its fine.

**Comment by harshavardhana** [comment]
We haven't migrated to go1.8 yet because of an existing problem go1.8 and Minio server - so this issue will be pending until then.

**Comment by harshavardhana** [comment]
https://github.com/minio/minio/pull/4203 - this PR should fix this issue as well. @vadmeste @minherz

**Comment by vadmeste** [comment]
@harshavardhana, stat() in your PR #4203 gives correct result:

```
stat(c:\volume) = &{volume {16 {2540406572 30589644} {2114449536 30589651} {2114449536 30589651} 0 0} 0  1855002566 393216 71690}, <nil>

os.Stat(c:\volume) = <nil>, GetFileAttributesEx \ContainerMappedDirectories\9600EE00-2FDB-4EBA-A836-99BBF3DF0DB9: The system cannot find the
 path specified.
```

:+1:

**Comment by harshavardhana** [comment]
> @harshavardhana, stat() in your PR #4203 gives correct result:
> 
> stat(c:\volume) = &{volume {16 {2540406572 30589644} {2114449536 30589651} {2114449536 30589651} 0 0} 0  1855002566 393216 71690}, <nil>
> 
> os.Stat(c:\volume) = <nil>, GetFileAttributesEx \ContainerMappedDirectories\9600EE00-2FDB-4EBA-A836-99BBF3DF0DB9: The system cannot find the
>  path specified.

Making relevant changes to address this issue as well.

**Comment by harshavardhana** [comment]
Fixed by #4203
